feat(client): add support for cloud storage configuration per project

leafty opened the pull request:
Fixes #2736.

/deploy renku=leafty/update-cypress-tests-cloud-storage renku-gateway=master renku-notebooks=master renku-data-services=main extra-values=dataService.image.pullPolicy=Always,global.gitlab.url=gitlab.dev.renku.ch,notebooks.cloudstorage.s3.enabled=true,amalthea.extraChildResources[0].group=com.ie.ibm.hpsys,amalthea.extraChildResources[0].name=datasets,notebooks.cloudstorage.s3.installDatashim=false

lorenzo-cavazzi commented:
I was wondering if we should use another sub-tab in the settings page just for the Cloud Storage instead of filling even more the sessions tab. That way, we won't need a modal to interact with single storages.

The proposed solution might work well but I feel there is a little too much content on the page. Perhaps we should consider adding white cards as in the other session page to separate the different parts better visually.

P.S. be sure to coordinate with @jdupre81 on this

leafty commented:
Some WIP for adding a new storage configuration:
![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/484883f8-016d-4bc9-b011-933e3dc8f10b)
![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/e8a58fba-d0c1-4eb7-978a-8f133b2595cb)

leafty commented:
Screenshots of this new feature:

### Anonymous users
![Screenshot from 2023-10-06 15-04-11](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/c2aa4231-7618-41b2-8a44-09eb8af3d5b8)

### User has no access
![Screenshot from 2023-10-06 15-04-39](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/508ef739-f257-4a5a-ae02-06c874d343d1)

### Variants depending on the support from `renku-notebooks`
![Screenshot from 2023-10-06 15-05-33](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/13f6097d-a0a3-4de1-b470-c16c807ec137)
![Screenshot from 2023-10-06 15-06-38](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/db7b304f-1ebb-4e71-a8c4-fd3f0bccaec1)
![Screenshot from 2023-10-06 15-06-57](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/b9df5000-d1e3-4c81-8c70-ffdff68f11c9)

### Add a new cloud storage
![Screenshot from 2023-10-06 15-08-17](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/2f5a20da-0d83-4263-a36a-acedc1b77520)
![Screenshot from 2023-10-06 15-07-34](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/747d033a-14f4-4943-94dc-e4b5380308e4)
![Screenshot from 2023-10-06 15-08-27](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/0f5d8ebb-6559-4615-a28b-60ab64fc1aa8)
![Screenshot from 2023-10-06 15-08-46](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/9b4b8704-e11f-4bc2-9467-67022065d7e7)
![Screenshot from 2023-10-06 15-22-52](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/9f525d15-84a5-4706-8aad-ef9a0c8d3f28)

### View cloud storage
![Screenshot from 2023-10-06 15-11-10](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/cabff174-2df2-4bc1-b46c-ddd11f52508f)

### Edit cloud storage
![Screenshot from 2023-10-06 15-12-14](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/86382136-c4fb-46b8-a12e-a3398907e14d)

### Cloud storage not supported
![Screenshot from 2023-10-06 15-12-58](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/e4075c98-9456-4b3b-a6e2-75e4ea910486)

### Cloud storage credentials
![Screenshot from 2023-10-06 15-13-18](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/1ddeded2-d5ae-4bff-9817-f425d809f47f)

### Quick start interrupted for credentials
![Screenshot from 2023-10-06 15-13-30](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/829430ab-f74d-425f-a032-1a49b270b692)

### Temporary cloud storage
![Screenshot from 2023-10-06 15-18-03](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/8c368f1e-31c3-42ec-87a0-87769ed799cc)
![Screenshot from 2023-10-06 15-17-39](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/e450222c-c9dc-47a5-88dc-1d4bd1309f01)

ciyer reviewed (changes requested):
I have reviewed all but 4 files. I'll need a little more time for the last four files, but I wanted to provide these comments now.

I like the code and the appearance in general! I have some suggestions for refactorings in a few places.

leafty commented on client/src/features/session/sessions.types.ts:
```diff
@@ -108,3 +108,26 @@ export interface PatchSessionParams {
+      readonly: boolean;
+      source_path: string;
+      target_path: string;
+    };
```
Updated in [53ea0b9](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/53ea0b9574dcaf5003f7058243df533b012773f4)

leafty commented on client/src/features/session/hooks/options/useDefaultCloudStorageOption.hook.ts:
```diff
@@ -0,0 +1,89 @@
+          : {}),
+        ...storage,
+      })
+    );
```
I don't want to lift this into a module function. Moved it to a function on the same file in [53ea0b9](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/53ea0b9574dcaf5003f7058243df533b012773f4).

leafty commented on client/src/features/project/utils/projectCloudStorage.utils.ts:
```diff
@@ -0,0 +1,80 @@
+
+  const providedSensitiveFields = Object.entries(configuration)
+    .filter(([, value]) => value === CLOUD_STORAGE_SENSITIVE_FIELD_TOKEN)
+    .map(([key]) => key);
```
Done in [53ea0b9](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/53ea0b9574dcaf5003f7058243df533b012773f4)

ciyer reviewed (changes requested):
Thanks for making those changes. I am testing the feature and I have some feedback on the UI.

Adding cloud storage to the project settings is quite intuitive to use. One small suggestion, could you add some info text explaining that the `name` is used to determine the mount point? 

<img width="786" alt="image" src="https://github.com/SwissDataScienceCenter/renku-ui/assets/1196411/c5ad071f-a900-4bf5-b2d3-d79c4586cbbb">

I wanted to add temporary storage to a session, but I have absolutely no idea what I need to enter into these fields:
<img width="786" alt="image" src="https://github.com/SwissDataScienceCenter/renku-ui/assets/1196411/85fe2c6d-9ef9-4561-9659-61d3673aa9bc">

Is this documented somewhere?

ciyer reviewed (changes requested):
I tested with with public and private AWS buckets, and when everything is right, it works very nicely. 🎉 

When there are problem, though, it is basically impossible to debug, since there is no useful feedback available in the UI. I realize that this may be out of scope for this issue, but I just wanted to provide the feedback, so that this could be improved on in the future.

One thing I noticed about private S3 buckets -- the UI autofilled the `access_key_id` and the `secret_access_key`. For the former, I think this is probably ok, but I would have expected the `secret_access_key` to be treated like a password and 1. not shown in plaintext and 2. not autofilled without me authorizing the action.

lokijuhy commented:
@leafty Re text copy in the screenshots:

I put suggestions for modified text in **bold**, and text removal in ~~strikethrough~~.

### Screenshot "Anonymous users"
  - "Only authenticated users can access cloud storage setting**s**."
  - "Log in to **view** **cloud storage** settings"

### Screenshots for Variants depending on the support from `renku-notebooks`
  - for all:
      - “… when starting local sessions from the CLI”. ← how about linking to the docs here? [renku session docs link](https://renku.readthedocs.io/en/latest/renku-python/docs/reference/commands/session.html) (or if we’re getting more specific docs for cloud storage, use those instead!)
  - Variant # 1:
      - "~~Note that~~ This instance of **RenkuLab** does not support…"
  - Variants # 2 & # 3:
      - “You can always mount cloud storage when starting local sessions with the CLI” - Do you mean “You can mount *other types of* cloud storage when starting local sessions with the CLI”?

### Screenshot for Temporary Cloud Storage dialog box
  - Add to the top of the box: “This cloud storage will be configured for this session only. To configure cloud storage permanently for this project, go to Project Settings.”

leafty commented:
> One thing I noticed about private S3 buckets -- the UI autofilled the `access_key_id` and the `secret_access_key`. For the former, I think this is probably ok, but I would have expected the `secret_access_key` to be treated like a password and 1. not shown in plaintext and 2. not autofilled without me authorizing the action.

I think I would like to address this in a subsequent PR. We have no input of type "password" at the moment and making a nice one is not trivial.

leafty commented on client/src/features/project/components/AddCloudStorageButton.tsx:
```diff
@@ -0,0 +1,862 @@
+
+          <div className="form-rk-green">
+            <div className="mb-3">
+              <Label className="form-label" for="addCloudStorageName">
```
Done in [b066797](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/b0667978e9aa57fae3043099379e8764be606f5f).

leafty commented on client/src/features/project/components/AddCloudStorageButton.tsx:
```diff
@@ -0,0 +1,862 @@
+            </p>
+
+            <div className="mb-3">
+              <Label className="form-label" for="addCloudStorageName">
```
Done in Done in [b066797](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/b0667978e9aa57fae3043099379e8764be606f5f).

leafty commented on client/src/features/project/components/AddCloudStorageButton.tsx:
```diff
@@ -0,0 +1,862 @@
+export default function AddCloudStorageButton() {
+  const [isOpen, setIsOpen] = useState(false);
+  const toggle = useCallback(() => {
+    setIsOpen((open) => !open);
```
Done in Done in [b066797](https://github.com/SwissDataScienceCenter/renku-ui/pull/2760/commits/b0667978e9aa57fae3043099379e8764be606f5f).
Made the modal backdrop static, i.e. the modal can only be closed by clicking on the close button or icon.

leafty commented:
## Screenshot "Anonymous users"

Updated:
![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/c1dd4a59-aa8c-493b-b362-7ef0aa8c19ec)

## Screenshots for Variants depending on the support from `renku-notebooks`

Message about local sessions removed, as the CLI cannot start local sessions (Docker) with cloud storage.
![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/fde8484f-7522-4692-8cc2-6b21f0fbdcb8)
![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/d946104a-4bdc-4094-bb8f-530daa376156)
![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/69118af0-17d9-4714-9177-32d48093b645)

## Screenshot for Temporary Cloud Storage dialog box

Added info box and explainer text about `rclone`.
![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/951086/a1f8733b-3195-4f99-a424-cffcec0407c4)

andre-code reviewed (changes requested):
Nice addition! I didn't conduct an exhaustive code review, as my colleagues have already done so, but I'd like to offer a couple of small suggestions to enhance the user experience:

1. Consider include an alert within the project session settings to inform users when they lack the necessary permissions, similar to how we handle it in the general project settings.

<img width="727" alt="Screenshot at Oct 11 17-08-59" src="https://github.com/SwissDataScienceCenter/renku-ui/assets/43388408/2eb83913-690c-43c4-8da8-099f3c46d7d0">

2. It might be beneficial to relocate the description for "rclone" to the input help label or position it below the "rclone" textarea for improved clarity.

<img width="728" alt="Screenshot at Oct 11 16-57-49" src="https://github.com/SwissDataScienceCenter/renku-ui/assets/43388408/e1d7eeef-bc93-49db-b748-64b2ee5b2ba0">

andre-code reviewed (approved):
🚀
